hs/src/LedgerState.hs

```diff
@@ -273,27 +270,39 @@ authTxin key txin (UTxO utxo') =
 -- than one witness.
 witnessed :: TxWits -> LedgerState -> Validity
 witnessed (TxWits tx wits) l =
-  if Set.size wits <= Set.size ins && all (hasWitness wits) ins
+  if neededSigners == signers && verified && notRedundant
     then Valid
-    else Invalid [InsufficientWitnesses]
+    else Invalid [IncorrectWitnesses]
   where
-    utxo'= l ^. utxoState . utxo
-    ins  = tx ^. inputs
-    hasWitness witnesses input =
-        isJust $ find (isWitness tx input utxo') witnesses
-    isWitness tx' input unspent (Wit key sig) =
-      verify key tx' sig && authTxin key input unspent
+    utxo' = l ^. utxoState . utxo
+    inputAuthors = Set.foldr insertHK Set.empty (tx ^. inputs)
```

Comment:
maybe extract this in an additional function to unclutter here

Reply:
done.